src/python/pants/engine/target.py

```diff
@@ -714,6 +714,36 @@ def expect_single(self) -> Target:
         return self[0]
 
 
+# If this predicate returns false, then the target's field's deps will be ignored.
+ShouldTraverseDepsPredicate: TypeAlias = (
+    "Callable[[Target, Dependencies | SpecialCasedDependencies], bool]"
+)
```

Comment:
How can we ensure that an arbitrary callable is correctly hashable?  I think functions will hash in Python but they can be closures with references to mutables, so I'm not sure this couldn't lead to subtle bugs.

Reply:
nice catch. quick test shows that the hash value for a closure changes also for the same inputs which suggests it uses the function address for the hash, so I don't think it matters what actual values are referenced by the closure but the hash value will most likely not be stable across runs let alone machines.

So, `ShouldTraverseDepsPredicate` would likely need some custom hash implementation..

Reply:
oh dang. Very true. I need to add a check somewhere (not sure where yet) that ensures `predicate.__closure__` is either not present or is `None`.

That also means I have to rework how I create the predicate in #19155 so that it is a hashable object with a `__call__` method that matches `ShouldTraverseDepsPredicate`.

Reply:
As far as the default implementations, `DepsTraversalPredicates.if_dependencies_field` and `DepsTraversalPredicates.always` are static functions that will always return the same hash. So, the default hash is sufficient there I think. This, however would be problematic, and I'm not sure how to protect against a lambda or other foot guns like this:

```python
def make_predicate() -> ShouldTraverseDepsPredicate:
    def predicate(target: Target, field: Dependencies | SpecialCasedDependencies) -> bool:
        return True
```

Reply:
How about enforcing the use of the predicate to be a subclass?, that removes quite a few guns from the equation:
```python

class ShouldTraverseDepsPredicate:
  def __call__(self, ...): ...

  def __hash__(): ...
```

then any custom predicates would have to subclass that one...

Reply:
Subclasses open up more things because you could add a reference to _anything_ under any name and we'd have to scan the `vars()` up the `__mro__` to know for sure it doesn't reference anything mutable.

Idk, this seems to open up a huge field of footguns since there's not really a way to vet that any callable object in Python doesn't have references to things that would make it unstable across runs (and machines).

Some of this perhaps could be vetted in testing, but I'm not sure how off the top of my head

Reply:
Does it really need to be an arbitrary predicate here?  Or can we craft a good list of potential ones?

Reply:
> Subclasses open up more things because you could add a reference to _anything_ under any name and we'd have to scan the `vars()` up the `__mro__` to know for sure it doesn't reference anything mutable.
> 
> Idk, this seems to open up a huge field of footguns since there's not really a way to vet that any callable object in Python doesn't have references to things that would make it unstable across runs (and machines).
> 
> Some of this perhaps could be vetted in testing, but I'm not sure how off the top of my head

This shouldn't be any worse than any other type we use (relying on the `__hash__()` method of the object). If you try to call `hash()` on something that is mutable you'll get a `TypeError` saying it's an unhashable type so dataclasses lends itself well for this purpose.

Reply:
For background: I designed this specifically for a case where I want a reference to something beyond target and field:
https://github.com/pantsbuild/pants/blob/c58ce117e2d99927811379a90de68fca49eda330/src/python/pants/core/goals/package.py#L187-L209

We already have dataclasses that include other dataclasses (eg one request that wraps another request), so the references to other things become more explicit. And subclassing avoids the issues with a closure because we would require that it be an instance of our class versus a random `Callable`. It also avoids the issues with the hash of a function where the hash is consistent within one pantsd session, but not across sessions.

I don't completely understand hashing and how long pants can cache rule Params for (args or results of the rule), but subclassing a frozen dataclass makes the most sense to me so far.

Reply:
(FYI when you subclass a frozen dataclass, you have to also declare it as a frozen dataclass, or else it gets unfrozen)

Reply:
Anyways, frozen dataclass subclass sounds good to me. And we should document this caveat about closures somewhere, because there's probably still some nonsense you could do with this no matter what we enforce, but at some point we can't stop all abuses

Reply:
🤷‍♂️ none of this is unique to this PR though. We use frozen dataclasses everywhere, so I don't think we have to do anything extra for safety

Unless there's something extra nasty I'm missing?

Reply:
I don't think we have many callable fields though. Maybe I'm over thinking this, but just seems prudent to at least avoid the closure footgun

Reply:
Also from reading the earlier comments, I think there might be some confusion as to caching, hashability, and mutability.

So, nothing is truly immutable in Python (unless you implement your stuff in C), you can only approximate immutability by making it really hard to mutate (frozen dataclasses are a good example. They can be mutated, but it takes gymnastics).

Hashability implicitly requires immutability because otherwise the hash wouldnt be useful. Even the stdlib approximates mutability by asking if the class is hashable.

Hashes of objects with similar data aren't guaranteed to be stable across process runs. So they are only good for in-process memoization.

We frequently use frozen dataclasses because they provide hash semantics for us (and provide "deep" hashing semantics).

Now enter Pants. At the rule boundaries, the inputs/outputs are cached in process (pantsd, the daemon, specifically). That's why the daemon gets you the big boost when issuing multiple pants commands. It's reusing the same process which has a warm cache of objects with hashes we assume are stable.

We don't use this cache across multiple processes (so restarts of the daemon, turning off the daemon, or across machines with remote caching, or remote execution), again because they aren't stable across processes.

The speedup you see across processes comes from caching `Digest`s (hashes of files/directory trees) and `Process`es (the object responsible for sandboxing and running a process.

So that's why we "require" immutability of rule input/outputs. If something isn't hashable, that'd straight up be a nonstarter. Closures are hashable, but you get a different hash per closure, so it wouldn't be incorrect, you'd just skip out on in-process caching, which is obviously not a great thing.

Now a closure itself is immutable, but like-closures generate different hashes so the caching isn't incorrect, it's just useless.

The issue is if it's referencing a mutable variable. So that's why only allowing a frozen dataclass is the safest and best way forward. It is itself immutable, and can only hold references to immutable members.

(Thanks again @jriddy for opening the issue. Good eye)

Reply:
Some exercises would be:
- trying to hash a frozen dataclass with a mutable member
- hashing a `frozenset` in multiple processes
- making a closure with a reference to a mutable object and seeing if it is hashable

Reply:
> (FYI when you subclass a frozen dataclass, you have to also declare it as a frozen dataclass, or else it gets unfrozen)

In my testing today, I was playing with a custom `__hash__` function on frozen dataclasses. But, it looks subclasses that are dataclasses _do not inherit generated functions like `__hash__`_. That's because the `@dataclass` decorator looks in the current class without traversing `__mro__` when determining if a method is defined. Then, if not defined in the current class, it adds it, effectively overwriting the parent `__hash__` method.

I was worried about that because `AsyncFieldMixin` and `Target` both have a custom `__hash__` function, but my worry was irrelevant because Field and Target don't use dataclasses.

The default predicate implementations are all logic and no data. The hash of any dataclass that has no fields is equivalent, which means that the cache key doesn't change when switching between `if_dependencies_field` and `always`. So, I have to use a custom hash that includes the `__class__` (or something), but that hash function is not inheritable (since each subclass has to have the `@dataclass(frozen=True)` decorator). So, I'm leaning away from dataclasses for this, similar to what `Target` and `AsyncFieldMixin` do.

Reply:
Also, a closure with a mutable object is hashable, but, and I don't think the implementation of function hashing takes the closure vars into account, I believe it just uses id like the default for object.

Reply:
Thanks for looking into that. All of that makes sense (btw the hash of a dataclass is supposed to be simple, like the hash of a tuple of the data).

So I think in an ideal world, the hash is the data plus the id of a function. Is there any way to have the types use that? A predicate and the data it needs to operate on

Reply:
Oh hey idea. 

Parent class is a frozen dataclass and an ABC where in `__init__` it stores the `__call__` attribute of the class (so `self._callable = type(self).__call__`). Then the parent has an abstract `__call__`.

Then children would also be frozen dataclasses, but the nice thing is their hash is reliant on the `__call__` function. Your "empty" predicate classes now have different hard because they have different `__call__` attributes.

Reply:
> (and also there's no point in making the parent class frozen because it has no members 😉)

besides enforcing subclasses to be frozen.. ;)

Reply:
> Oh hey idea.
> 
> Parent class is a frozen dataclass and an ABC where in `__init__` it stores the `__call__` attribute of the class (so `self._callable = type(self).__call__`. Then the parent has an abstract `__call__`.
> 
> Then children would also be frozen dataclasses, but the nice thing is their hash is reliant on the `__call__` function. Your "empty" predicate classes now have different hard because they have different `__call__` attributes.

Thanks @thejcannon! This worked very well. I had to tweak it a bit to avoid issues when the subclass is also decorated with `@dataclass(frozen=True)` because `__init__` gets re-generated without regard to `__init__` being defined in the parent class.; I moved populating `self._callable` to `__post_init__`.